#10141 : Document::location set null for documents without a browsing context

slayerjain opened the pull request:
Fixes #10141.

<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/10257)

<!-- Reviewable:end -->

slayerjain committed:
issue #10141 : Document::location set null for documents without a browsing context. r=Ms2ger

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @simonsapin (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @kichjang: components/script/dom/webidls/Document.webidl, components/script/dom/document.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

Ms2ger commented:
Did you compile this locally?

slayerjain commented:
oh.. my bad! switching my laptop created the mess!

Ms2ger commented:
No problem; let me know when it works :)

slayerjain commented:
The internet bandwidth in my country is really pathetic! couldn't build servo yet!

slayerjain commented:
I've fixed my laptop's repo,  but now I'm stuck with an error. I've also modified xmldocument.rs's location method to return Option<Root<Location>>
<img width="1236" alt="screen shot 2016-03-30 at 10 53 22 am" src="https://cloud.githubusercontent.com/assets/12831254/14132644/ca0130de-f665-11e5-9f94-0406130922a1.png">

Ms2ger commented:
Ah, I forgot about that. Attributes that can return null are called `GetFoo` rather than `Foo`, so you'll need to rename those two implementations.

nox commented:
@slayerjain Please rebase your commit on top of master. There shouldn't be any merge commit in this branch. Feel free to ask questions if you don't know how to do this, either here or on IRC.

slayerjain commented:
@nox Can you please explain the process, just to verify what i'm thinking? :)

nox commented:
@slayerjain There needs to be only one commit on the branch, https://github.com/servo/servo/pull/10257/commits/04f648805d1b7533f0e4881967378d3e05146b1f, and there shouldn't be any merge on top of it.

So I update my local master branch from upstream, check out the branch first_bug, do `git rebase -i master`, remove the second line, save the file and quit, and let Git do its magic.

KiChjang commented:
Here's [documentation](https://github.com/servo/servo/wiki/Github-workflow) on how the entire workflow works.

slayerjain commented:
@nox Awesome! on it! :)

slayerjain commented:
@nox done! :)

KiChjang commented on components/script/dom/document.rs:
```diff
@@ -2359,8 +2359,11 @@ impl DocumentMethods for Document {
+        match self.browsing_context() {
+            Some(browsingcontext) => return Some(self.location.or_init(|| Location::new(&self.window))),
+            None => return None,
+        }
```
This could be
```rust
self.browsing_context().and_then(|_| {
    Some(self.location.or_init(|| Location::new(&self.window))
})
```

nox replied:
AFAICT it can just be:

```rust
self.browsing_context().map(|_| self.location.or_init(|| Location::new(&self.window)))
```

KiChjang replied:
Ugh. I still can't tell when it's okay to use `.map()`.

frewsxcv replied:
Lame attempt at clarifying:

If you want the variant to go from `Some` → `Some` use `map`

If you want the variant to go from `Some` → (`Some` or `None`) use `and_then`

slayerjain commented:
@KiChjang @nox noted! :)

slayerjain commented:
@Ms2ger can you please review it? :)

Ms2ger commented:
Code looks good, but I hope there's tests for this. Let's ask the bot…

@bors-servo try

---

Reviewed 1 of 4 files at r2, 2 of 3 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/servo/10257#-:-KEl639JZARvGyB5NWG1:2128449934)*
<!-- Sent from Reviewable.io -->

bors-servo commented:
:hourglass: Trying commit 097f822 with merge e0651b7...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/169)

Ms2ger commented:
I made https://github.com/slayerjain/servo/pull/1 for the test failures; please squash that into your commit and r=me.

slayerjain commented:
why did these tests fail now? :/

KiChjang commented:
Don't worry about the failures; it's been fixed in a recent PR. More importantly, I'll have to ask you to squash your commits into one.

slayerjain commented:
@KiChjang Thanks! done. :)

KiChjang commented:
@bors-servo r+

Thanks!

bors-servo commented:
:pushpin: Commit 4de04bb has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang 4de04bb8be0e030792809b78bb5e3113ce16b1fc -->

KiChjang commented:
@bors-servo r=Ms2ger

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #10494

bors-servo commented:
:pushpin: Commit 4de04bb has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger 4de04bb8be0e030792809b78bb5e3113ce16b1fc -->

slayerjain commented:
what does this "@bors-servo r=Ms2ger" do ?

bors-servo commented:
:hourglass: Testing commit 4de04bb with merge 59d205c...

KiChjang commented:
It's just telling our buildbot to accept this PR on behalf of Ms2ger.

slayerjain commented:
Didn't know that was possible, feels kinda odd.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/199)

KiChjang commented:
```
Ran 4652 tests finished in 1073.0 seconds.
  • 4647 ran as expected. 1361 tests skipped.
  • 5 tests had unexpected subtest results

Tests with unexpected results:
  ▶ Unexpected subtest result in /XMLHttpRequest/responsexml-document-properties.htm:
  └ PASS [expected FAIL] location

  ▶ Unexpected subtest result in /domparsing/DOMParser-parseFromString-html.html:
  └ PASS [expected FAIL] Location value

  ▶ Unexpected subtest result in /domparsing/DOMParser-parseFromString-xml.html:
  └ PASS [expected FAIL] Should parse correctly in type text/xml

  ▶ Unexpected subtest result in /html/browsers/history/the-location-interface/document_location.html:
  └ PASS [expected FAIL] document not in a browsing context

  ▶ Unexpected subtest result in /_mozilla/mozilla/windowproxy.html:
  └ PASS [expected FAIL] document.location is the right thing on non-rendered document
```
Looks like you'll need to update more test expectations.

slayerjain commented:
@KiChjang where to make the changes?

KiChjang commented:
Wait, you didn't squash your changes, you REMOVED his commit.

His commit had the diff and the files that you can use to update the test expectations, see slayerjain/servo#1 for where to edit them.

slayerjain commented:
@bors-servo try

slayerjain commented:
i thought servo bot would run some tests.. :/

KiChjang commented:
Unfortunately, bors-servo only listens to people with try access or reviewers. Sorry!

@bors-servo r+

bors-servo commented:
:pushpin: Commit 6626c5c has been approved by `KiChjang`

<!-- @bors-servo r=KiChjang 6626c5cfd492239aa716438a3b9d6b2e94999956 -->

bors-servo commented:
:hourglass: Testing commit 6626c5c with merge c33bf49...

slayerjain commented:
@Ms2ger can't wait to use it :)

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/191), [arm32](http://build.servo.org/builders/arm32/builds/207), [arm64](http://build.servo.org/builders/arm64/builds/199), [linux-dev](http://build.servo.org/builders/linux-dev/builds/203), [linux-rel](http://build.servo.org/builders/linux-rel/builds/221), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/208), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/201), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/205)
